### ISSUE TITLE
"if not" vs "unless"

### ISSUE THREAD

**Issue by giddie** [rubocop-hq/ruby-style-guide#329]
Maybe a question more than an issue: I've gravitated toward using `if not` in certain circumstances, and I've figured out why:

In situations where I expect the condition to be false, and the code will normally run, I use `unless`:

``` ruby
unless name.empty?
  puts "Hello #{name}"
end
```

However, whenever I expect the condition to be true, and the code will _not_ normally run, I use `if not`:

``` ruby
if not name == 'Paul'
  STDERR.puts "Unauthorized!"
end
```

Most people would probably use `unless` across the board, but in the latter example I'd find that difficult to understand at a glance, simply because the words "if" and "unless" have this semantic meaning associated with them. I'm interested in thoughts on this. This isn't something I particularly planned, so it's really not an attempt to over-complicate: I fell into this pattern of use because of my brain finding it easier to grok the keywords this way, and I eventually realised what I was doing. Thoughts?

**Comment by strand** [comment]
Would a good rule here be "Use `if !` if the condition is expected to be false, and `unless` if the condition is likely to be true." ? Does that phrasing even make sense?

**Comment by Nowaker** [comment]
The whole argument doesn't make sense as the given example is `if not name == 'Paul'` whereas the only reasonable way to write it is `if name != 'Paul'`. 

In a referenced #325 I said:

> `do something if negative_expression` > `do something unless positive_expression`
> `raise if array.empty?` > `raise unless array.full?`
>
>"If you can use if without not [negation], prefer it to unless."
>
> And this doesn't have anything to do with condition being rare or common.

...or your expectation for the expression to be true or false either.

**Comment by cheerfulstoic** [comment]
@strand I like your suggestion of phrasing.  The only thing that I would say is that in my experience most of the time you'll expect either a `true` or a `false` result most of the time, but sometimes it's not clear how often it will go one way or the other.  In those cases I think it's fine to prioritize making the condition positive and adjusting the `if` / `unless` accordingly

@Nowaker I'm not exactly sure what you're saying, but I agree that `if name != 'Paul'` > `if not name == 'Paul'`.  I think if the method can be changed to express negation (be that `#==`/`#!=` or `#valid?`/`#not_valid?`) that that is preferable.

Actually, I think I understand what you're saying now, but I'll leave what I said ;)  But yeah, I think rareness only comes into play when you only have a positive expression/method available to you.

**Comment by giddie** [comment]
I like @strand's phrasing, but I think it's the opposite way round, isn't it?

```ruby
# using if !
raise_alarm!  if !access_permitted?
do_something  if !expected_condition?

# using unless
open_door!  unless locked?
do_something  unless unexpected_condition?
```
So: `if !` if the condition is _likely_, and `unless` if the condition is _unlikely_.

@Nowaker Also agree. In the OP, I suggested `if not`, but since then I have been shown that this is a bad use of the `not` keyword. I also agree that the condition itself should ideally handle the negation, and my original example was not very good for that reason. I avoided comparison operators in later examples for that reason.

**Comment by dreyks** [comment]
Sorry for a necropost and this may be an off-topic too, but this discussion just helped me to formalize the things in my head about another conditional construct "and/or".

If talking about Englishness and code understanding I tend to like
```ruby
all_is_good? or return

some_variable = get_some_variable or raise
```
and not to particularly like
```ruby
something_failed? and return
```
In other words I tend to replace the post-conditional `unless` for a `or`, but bot so much an `if` into an `and`.
I understand that this may be not a proper place to discuss it, moreover this has almost definitely nothing to do with the automated check tools, cause they won't understand it. `disregard_this_text if so` (SWIDT)

**Comment by meagar** [comment]
@drewda All of those are unnecessary. Avoid `and`/`or` to begin with and just use postfix `if`/`unless`

Use `return if something_bad?` and `raise ... unless all_is_good?` instead of `something_bad? and return` and `all_is_good? or raise ...`

**Comment by cheerfulstoic** [comment]
Also, to clarify, when having local operators is appropriate `&&` and `||` are better.  See:

https://github.com/bbatsov/ruby-style-guide#no-and-or-or

Though I think there are still some cases where `and` and `or` make sense.  Avdi Grimm's screencast matches my thoughts:

http://devblog.avdi.org/2014/08/26/how-to-use-rubys-english-andor-operators-without-going-nuts/

**Comment by dreyks** [comment]
Yes, Avdi's blog is where it all started. He seems to use only the perl-style `or` though
